### DISCUSSION TITLE
PDF compression

### DISCUSSION THREAD

**Discussion by lbr991** [pymupdf/PyMuPDF#2782] [Looking for help]
https://pymupdf.readthedocs.io/en/latest/document.html#Document.save
What is the difference between `garbage=3` and `garbage=4`? 
If time delay is not an issue and max possible compression is the objective, should `garbage=4` be used? 

Also for max compression, should I just use `deflate=True` or also add `deflate_images=True` and `deflate_fonts=True`? Is it good practice?

**Comment by JorjMcKie** [comment]
> Also for max compression, should I just use deflate=True or also add deflate_images=True and deflate_fonts=True? Is it good practice?

At least it is no problem! But deflate=True will switch on the other two.

Garbage option 3 searches all PDF objects (xrefs) and compares their definitions. If equal ones are found, one of the two is deleted and references to it are changed to point to the remaining one.

Garbage option 4 in addition looks at so-called "streams" which is the PDF term for large data portions - usually binary content like images or fonts, but could be anything.
If equality is found, the same procedure is used: drop one, live with the other one.
This options obviously has a high potential for size reduction, but it does take longer understandably.

Garbage collection options are arranged in a hierarchy: option `n` does everything what any option `m < n` does.

**Comment by JorjMcKie** [comment]
You may like to hear, that we intend to extend compression options with the next release of our base library MuPDF in spring of next year.
